RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc

```diff
@@ -968,6 +977,32 @@ void MuonIdProducer::fillMuonId(edm::Event& iEvent, const edm::EventSetup& iSetu
    // fillTime( iEvent, iSetup, aMuon );
 }
 
+void MuonIdProducer::arbitrateMuons( reco::MuonCollection* muons, reco::CaloMuonCollection* caloMuons )
+{
+  reco::Muon::ArbitrationType arbitration = reco::Muon::SegmentAndTrackArbitration;
+  // arbitrate TrackerMuons
+  // if a muon was exclusively TrackerMuon check if it can be a calo muon
+  for (reco::MuonCollection::iterator muon=muons->begin(); muon!=muons->end();){
+    if (muon->isTrackerMuon()){
+      if (muon->numberOfMatches(arbitration)<minNumberOfMatches_){
+	// TrackerMuon failed arbitration
+	// If not any other base type - erase the element
+	// (PFMuon is not a base type)
+	unsigned int mask = reco::Muon::TrackerMuon|reco::Muon::PFMuon;
+	if ((muon->type() & (~mask))==0){
+	  const reco::CaloMuon& caloMuon = makeCaloMuon(*muon);
+	  if (isGoodCaloMuon(caloMuon)) caloMuons->push_back( caloMuon );
+	  muon = muons->erase(muon);
+	  continue;
+	} else {
+	  muon->setType(muon->type()&(~reco::Muon::TrackerMuon));
+	}
+      } 
+    }
+    muon++;
```

Comment:
erase happens on the last element, then in ``` muon = muons->erase(muon); ```  the ```muon``` is the same as the end() and in this line you will apply end()++, which is ill-defined and will not match the for loop condition